applications/mqtt/mosquitto-1.3.4/man/mosquitto.conf.5
'\" t
.\"     Title: mosquitto.conf
.\"    Author: [see the "Author" section]
.\" Generator: DocBook XSL Stylesheets v1.78.1 <http://docbook.sf.net/>
.\"      Date: 08/28/2014
.\"    Manual: File formats and conventions
.\"    Source: Mosquitto Project
.\"  Language: English
.\"
.TH "MOSQUITTO\&.CONF" "5" "08/28/2014" "Mosquitto Project" "File formats and conventions"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
mosquitto.conf \- the configuration file for mosquitto
.SH "SYNOPSIS"
.HP \w'\fBmosquitto\&.conf\fR\ 'u
\fBmosquitto\&.conf\fR
.SH "DESCRIPTION"
.PP
\fBmosquitto\&.conf\fR
is the configuration file for mosquitto\&. This file can reside anywhere as long as mosquitto can read it\&. By default, mosquitto does not need a configuration file and will use the default values listed below\&. See
\fBmosquitto\fR(8)
for information on how to load a configuration file\&.
.SH "FILE FORMAT"
.PP
All lines with a # as the very first character are treated as a comment\&.
.PP
Configuration lines start with a variable name\&. The variable value is separated from the name by a single space\&.
.SH "AUTHENTICATION"
.PP
The authentication options described below allow a wide range of possibilities in conjunction with the listener options\&. This section aims to clarify the possibilities\&.
.PP
The simplest option is to have no authentication at all\&. This is the default if no other options are given\&. Unauthenticated encrypted support is provided by using the certificate based SSL/TLS based options cafile/capath, certfile and keyfile\&.
.PP
MQTT provides username/password authentication as part of the protocol\&. Use the password_file option to define the valid usernames and passwords\&. Be sure to use network encryption if you are using this option otherwise the username and password will be vulnerable to interception\&.
.PP
When using certificate based encryption there are two options that affect authentication\&. The first is require_certificate, which may be set to true or false\&. If false, the SSL/TLS component of the client will verify the server but there is no requirement for the client to provide anything for the server: authentication is limited to the MQTT built in username/password\&. If require_certificate is true, the client must provide a valid certificate in order to connect successfully\&. In this case, the second option, use_identity_as_username, becomes relevant\&. If set to true, the Common Name (CN) from the client certificate is used instead of the MQTT username for access control purposes\&. The password is not replaced because it is assumed that only authenticated clients have valid certificates\&. If use_identity_as_username is false, the client must authenticate as normal (if required by password_file) through the MQTT options\&.
.PP
When using pre\-shared\-key based encryption through the psk_hint and psk_file options, the client must provide a valid identity and key in order to connect to the broker before any MQTT communication takes place\&. If use_identity_as_username is true, the PSK identity is used instead of the MQTT username for access control purposes\&. If use_identity_as_username is false, the client may still authenticate using the MQTT username/password if using the password_file option\&.
.PP
Both certificate and PSK based encryption are configured on a per\-listener basis\&.
.PP
Authentication plugins can be created to replace the password_file and psk_file options (as well as the ACL options) with e\&.g\&. SQL based lookups\&.
.PP
It is possible to support multiple authentication schemes at once\&. A config could be created that had a listener for all of the different encryption options described above and hence a large number of ways of authenticating\&.
.SH "GENERAL OPTIONS"
.PP
\fBacl_file\fR \fIfile path\fR
.RS 4
Set the path to an access control list file\&. If defined, the contents of the file are used to control client access to topics on the broker\&.
.sp
If this parameter is defined then only the topics listed will have access\&. Topic access is added with lines of the format:
.sp
topic [read|write] <topic>
.sp
The access type is controlled using "read" or "write"\&. This parameter is optional \- if not given then the access is read/write\&. <topic> can contain the + or # wildcards as in subscriptions\&.
.sp
The first set of topics are applied to anonymous clients, assuming
\fBallow_anonymous\fR
is true\&. User specific topic ACLs are added after a user line as follows:
.sp
user <username>
.sp
The username referred to here is the same as in
\fBpassword_fil\fRe\&. It is not the clientid\&.
.sp
It is also possible to define ACLs based on pattern substitution within the topic\&. The form is the same as for the topic keyword, but using pattern as the keyword\&.
.sp
pattern [read|write] <topic>
.sp
The patterns available for substition are:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
%c to match the client id of the client
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
%u to match the username of the client
.RE
.sp
The substitution pattern must be the only text for that level of hierarchy\&. Pattern ACLs apply to all users even if the "user" keyword has previously been given\&.
.sp
Example:
.sp
pattern write sensor/%u/data
.sp
Allow access for bridge connection messages:
.sp
pattern write $SYS/broker/connection/%c/state
.sp
If the first character of a line of the ACL file is a # it is treated as a comment\&.
.sp
Reloaded on reload signal\&. The currently loaded ACLs will be freed and reloaded\&. Existing subscriptions will be affected after the reload\&.
.RE
.PP
\fBallow_anonymous\fR [ true | false ]
.RS 4
Boolean value that determines whether clients that connect without providing a username are allowed to connect\&. If set to
\fIfalse\fR
then another means of connection should be created to control authenticated client access\&. Defaults to
\fItrue\fR\&.
.sp
Reloaded on reload signal\&.
.RE
.PP
\fBallow_duplicate_messages\fR [ true | false ]
.RS 4
If a client is subscribed to multiple subscriptions that overlap, e\&.g\&. foo/# and foo/+/baz , then MQTT expects that when the broker receives a message on a topic that matches both subscriptions, such as foo/bar/baz, then the client should only receive the message once\&.
.sp
Mosquitto keeps track of which clients a message has been sent to in order to meet this requirement\&. This option allows this behaviour to be disabled, which may be useful if you have a large number of clients subscribed to the same set of topics and want to minimise memory usage\&.
.sp
It can be safely set to
\fItrue\fR
if you know in advance that your clients will never have overlapping subscriptions, otherwise your clients must be able to correctly deal with duplicate messages even when then have QoS=2\&.
.sp
Defaults to
\fItrue\fR\&.
.sp
Reloaded on reload signal\&.
.RE
.PP
\fBauth_opt_*\fR \fIvalue\fR
.RS 4
Options to be passed to the auth plugin\&. See the specific plugin instructions\&.
.RE
.PP
\fBauth_plugin\fR \fIfile path\fR
.RS 4
Specify an external module to use for authentication and access control\&. This allows custom username/password and access control functions to be created\&.
.sp
Not currently reloaded on reload signal\&.
.RE
.PP
\fBautosave_interval\fR \fIseconds\fR
.RS 4
The number of seconds that mosquitto will wait between each time it saves the in\-memory database to disk\&. If set to 0, the in\-memory database will only be saved when mosquitto exits or when receiving the SIGUSR1 signal\&. Note that this setting only has an effect if persistence is enabled\&. Defaults to 1800 seconds (30 minutes)\&.
.sp
Reloaded on reload signal\&.
.RE
.PP
\fBautosave_on_changes\fR [ true | false ]
.RS 4
If
\fItrue\fR, mosquitto will count the number of subscription changes, retained messages received and queued messages and if the total exceeds
\fBautosave_interval\fR
then the in\-memory database will be saved to disk\&. If
\fIfalse\fR, mosquitto will save the in\-memory database to disk by treating
\fBautosave_interval\fR
as a time in seconds\&.
.sp
Reloaded on reload signal\&.
.RE
.PP
\fBclientid_prefixes\fR \fIprefix\fR
.RS 4
If defined, only clients that have a clientid with a prefix that matches clientid_prefixes will be allowed to connect to the broker\&. For example, setting "secure\-" here would mean a client "secure\-client" could connect but another with clientid "mqtt" couldn\*(Aqt\&. By default, all client ids are valid\&.
.sp
Reloaded on reload signal\&. Note that currently connected clients will be unaffected by any changes\&.
.RE
.PP
\fBconnection_messages\fR [ true | false ]
.RS 4
If set to
\fItrue\fR, the log will include entries when clients connect and disconnect\&. If set to
\fIfalse\fR, these entries will not appear\&.
.sp
Reloaded on reload signal\&.
.RE
.PP
\fBinclude_dir\fR \fIdir\fR
.RS 4
External configuration files may be included by using the include_dir option\&. This defines a directory that will be searched for config files\&. All files that end in \*(Aq\&.conf\*(Aq will be loaded as a configuration file\&. It is best to have this as the last option in the main file\&. This option will only be processed from the main configuration file\&. The directory specified must not contain the main configuration file\&.
.RE
.PP
\fBlog_dest\fR \fIdestinations\fR
.RS 4
Send log messages to a particular destination\&. Possible destinations are:
\fBstdout\fR\fBstderr\fR\fBsyslog\fR\fBtopic\fR\&.
.sp
\fBstdout\fR
and
\fBstderr\fR
log to the console on the named output\&.
.sp
\fBsyslog\fR
uses the userspace syslog facility which usually ends up in /var/log/messages or similar and topic logs to the broker topic \*(Aq$SYS/broker/log/<severity>\*(Aq, where severity is one of D, E, W, N, I, M which are debug, error, warning, notice, information and message\&. Message type severity is used by the subscribe and unsubscribe log_type options and publishes log messages at $SYS/broker/log/M/subscribe and $SYS/broker/log/M/unsubscribe\&.
.sp
The
\fBfile\fR
destination requires an additional parameter which is the file to be logged to, e\&.g\&. "log_dest file /var/log/mosquitto\&.log"\&. The file will be closed and reopened when the broker receives a HUP signal\&. Only a single file destination may be configured\&.
.sp
Use "log_dest none" if you wish to disable logging\&. Defaults to stderr\&. This option may be specified multiple times\&.
.sp
Note that if the broker is running as a Windows service it will default to "log_dest none" and neither stdout nor stderr logging is available\&.
.sp
Reloaded on reload signal\&.
.RE
.PP
\fBlog_timestamp\fR [ true | false ]
.RS 4
Boolean value, if set to
\fItrue\fR
a timestamp value will be added to each log entry\&. The default is
\fItrue\fR\&.
.sp
Reloaded on reload signal\&.
.RE
.PP
\fBlog_type\fR \fItypes\fR
.RS 4
Choose types of messages to log\&. Possible types are:
\fIdebug\fR,
\fIerror\fR,
\fIwarning\fR,
\fInotice\fR,
\fIinformation\fR,
\fInone\fR,
\fIall\fR\&. Defaults to
\fIerror\fR,
\fIwarning\fR,
\fInotice \fRand
\fIinformation\fR\&. This option may be specified multiple times\&. Note that the
\fIdebug \fRtype (used for decoding incoming/outgoing network packets) is never logged in topics\&.
.sp
Reloaded on reload signal\&.
.RE
.PP
\fBmax_inflight_messages\fR \fIcount\fR
.RS 4
The maximum number of QoS 1 or 2 messages that can be in the process of being transmitted simultaneously\&. This includes messages currently going through handshakes and messages that are being retried\&. Defaults to 20\&. Set to 0 for no maximum\&. If set to 1, this will guarantee in\-order delivery of messages\&.
.sp
Reloaded on reload signal\&.
.RE
.PP
\fBmax_queued_messages\fR \fIcount\fR
.RS 4
The maximum number of QoS 1 or 2 messages to hold in the queue above those messages that are currently in flight\&. Defaults to 100\&. Set to 0 for no maximum (not recommended)\&. See also the
\fBqueue_qos0_messages\fR
option\&.
.sp
Reloaded on reload signal\&.
.RE
.PP
\fBmessage_size_limit\fR \fIlimit\fR
.RS 4
This option sets the maximum publish payload size that the broker will allow\&. Received messages that exceed this size will not be accepted by the broker\&. The default value is 0, which means that all valid MQTT messages are accepted\&. MQTT imposes a maximum payload size of 268435455 bytes\&.
.RE
.PP
\fBpassword_file\fR \fIfile path\fR
.RS 4
Set the path to a password file\&. If defined, the contents of the file are used to control client access to the broker\&. The file can be created using the
\fBmosquitto_passwd\fR(1)
utility\&. If mosquitto is compiled without TLS support (it is recommended that TLS support is included), then the password file should be a text file with each line in the format "username:password", where the colon and password are optional but recommended\&. If
\fBallow_anonymous\fR
is set to
\fIfalse\fR, only users defined in this file will be able to connect\&. Setting
\fBallow_anonymous\fR
to
\fItrue\fR
when
\fIpassword_file\fRis defined is valid and could be used with acl_file to have e\&.g\&. read only guest/anonymous accounts and defined users that can publish\&.
.sp
Reloaded on reload signal\&. The currently loaded username and password data will be freed and reloaded\&. Clients that are already connected will not be affected\&.
.sp
See also
\fBmosquitto_passwd\fR(1)\&.
.RE
.PP
\fBpersistence\fR [ true | false ]
.RS 4
If
\fItrue\fR, connection, subscription and message data will be written to the disk in mosquitto\&.db at the location dictated by persistence_location\&. When mosquitto is restarted, it will reload the information stored in mosquitto\&.db\&. The data will be written to disk when mosquitto closes and also at periodic intervals as defined by autosave_interval\&. Writing of the persistence database may also be forced by sending mosquitto the SIGUSR1 signal\&. If
\fIfalse\fR, the data will be stored in memory only\&. Defaults to
\fIfalse\fR\&.
.sp
Reloaded on reload signal\&.
.RE
.PP
\fBpersistence_file\fR \fIfile name\fR
.RS 4
The filename to use for the persistent database\&. Defaults to mosquitto\&.db\&.
.sp
Reloaded on reload signal\&.
.RE
.PP
\fBpersistence_location\fR \fIpath\fR
.RS 4
The path where the persistence database should be stored\&. Must end in a trailing slash\&. If not given, then the current directory is used\&.
.sp
Reloaded on reload signal\&.
.RE
.PP
\fBpersistent_client_expiration\fR \fIduration\fR
.RS 4
This option allows persistent clients (those with clean session set to false) to be removed if they do not reconnect within a certain time frame\&. This is a non\-standard option\&. As far as the MQTT spec is concerned, persistent clients persist forever\&.
.sp
Badly designed clients may set clean session to false whilst using a randomly generated client id\&. This leads to persistent clients that will never reconnect\&. This option allows these clients to be removed\&.
.sp
The expiration period should be an integer followed by one of d w m y for day, week, month and year respectively\&. For example:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
persistent_client_expiration 2m
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
persistent_client_expiration 14d
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
persistent_client_expiration 1y
.RE
.sp
As this is a non\-standard option, the default if not set is to never expire persistent clients\&.
.sp
Reloaded on reload signal\&.
.RE
.PP
\fBpid_file\fR \fIfile path\fR
.RS 4
Write a pid file to the file specified\&. If not given (the default), no pid file will be written\&. If the pid file cannot be written, mosquitto will exit\&. This option only has an effect is mosquitto is run in daemon mode\&.
.sp
If mosquitto is being automatically started by an init script it will usually be required to write a pid file\&. This should then be configured as e\&.g\&. /var/run/mosquitto\&.pid
.sp
Not reloaded on reload signal\&.
.RE
.PP
\fBpsk_file\fR \fIfile path\fR
.RS 4
Set the path to a pre\-shared\-key file\&. This option requires a listener to be have PSK support enabled\&. If defined, the contents of the file are used to control client access to the broker\&. Each line should be in the format "identity:key", where the key is a hexadecimal string with no leading "0x"\&. A client connecting to a listener that has PSK support enabled must provide a matching identity and PSK to allow the encrypted connection to proceed\&.
.sp
Reloaded on reload signal\&. The currently loaded identity and key data will be freed and reloaded\&. Clients that are already connected will not be affected\&.
.RE
.PP
\fBqueue_qos0_messages\fR [ true | false ]
.RS 4
Set to
\fItrue\fR
to queue messages with QoS 0 when a persistent client is disconnected\&. These messages are included in the limit imposed by max_queued_messages\&. Defaults to
\fIfalse\fR\&.
.sp
Note that the MQTT v3\&.1 spec states that only QoS 1 and 2 messages should be saved in this situation so this is a non\-standard option\&.
.sp
Reloaded on reload signal\&.
.RE
.PP
\fBretained_persistence\fR [ true | false ]
.RS 4
This is a synonym of the
\fBpersistence\fR
option\&.
.sp
Reloaded on reload signal\&.
.RE
.PP
\fBretry_interval\fR \fIseconds\fR
.RS 4
The integer number of seconds after a QoS=1 or QoS=2 message has been sent that mosquitto will wait before retrying when no response is received\&. If unset, defaults to 20 seconds\&.
.sp
Reloaded on reload signal\&.
.RE
.PP
\fBstore_clean_interval\fR \fIseconds\fR
.RS 4
The integer number of seconds between the internal message store being cleaned of messages that are no longer referenced\&. Lower values will result in lower memory usage but more processor time, higher values will have the opposite effect\&. Setting a value of 0 means the unreferenced messages will be disposed of as quickly as possible\&. Defaults to 10 seconds\&.
.sp
Reloaded on reload signal\&.
.RE
.PP
\fBsys_interval\fR \fIseconds\fR
.RS 4
The integer number of seconds between updates of the $SYS subscription hierarchy, which provides status information about the broker\&. If unset, defaults to 10 seconds\&.
.sp
Set to 0 to disable publishing the $SYS hierarchy completely\&.
.sp
Reloaded on reload signal\&.
.RE
.PP
\fBupgrade_outgoing_qos\fR [ true | false ]
.RS 4
The MQTT specification requires that the QoS of a message delivered to a subscriber is never upgraded to match the QoS of the subscription\&. Enabling this option changes this behaviour\&. If
\fBupgrade_outgoing_qos\fR
is set
\fItrue\fR, messages sent to a subscriber will always match the QoS of its subscription\&. This is a non\-standard option not provided for by the spec\&. Defaults to
\fIfalse\fR\&.
.sp
Reloaded on reload signal\&.
.RE
.PP
\fBuser\fR \fIusername\fR
.RS 4
When run as root, change to this user and its primary group on startup\&. If mosquitto is unable to change to this user and group, it will exit with an error\&. The user specified must have read/write access to the persistence database if it is to be written\&. If run as a non\-root user, this setting has no effect\&. Defaults to mosquitto\&.
.sp
This setting has no effect on Windows and so you should run mosquitto as the user you wish it to run as\&.
.sp
Not reloaded on reload signal\&.
.RE
.SH "LISTENERS"
.PP
The network ports that mosquitto listens on can be controlled using listeners\&. The default listener options can be overridden and further listeners can be created\&.
.SS "General Options"
.PP
\fBbind_address\fR \fIaddress\fR
.RS 4

applications/mqtt/mosquitto-1.3.4/src/mosquitto.c
/*
Copyright (c) 2009-2013 Roger Light <roger@atchoo.org>
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice,
   this list of conditions and the following disclaimer.
2. Redistributions in binary form must reproduce the above copyright
   notice, this list of conditions and the following disclaimer in the
   documentation and/or other materials provided with the distribution.
3. Neither the name of mosquitto nor the names of its
   contributors may be used to endorse or promote products derived from
   this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
POSSIBILITY OF SUCH DAMAGE.
*/

#include <config.h>

#ifndef WIN32
/* For initgroups() */
#  define _BSD_SOURCE
#  include <unistd.h>
#  include <grp.h>
#endif

#ifndef WIN32
#include <pwd.h>
#else
#include <process.h>
#include <winsock2.h>
#include <ws2tcpip.h>
#endif

#include <errno.h>
#include <signal.h>
#include <stdio.h>
#include <string.h>

// FIXME: started getting compilation warnings (implicit declarations for
// functions gettimeofday() and time(). temporarily solved like this.
#include <sys/time.h>

#ifdef WITH_WRAP
#include <tcpd.h>
#endif

#include "mosquitto_broker.h"
#include "memory_mosq.h"
#include "util_mosq.h"

struct mosquitto_db int_db;

bool flag_reload = false;
#ifdef WITH_PERSISTENCE
bool flag_db_backup = false;
#endif
bool flag_tree_print = false;
int run;
#ifdef WITH_WRAP
#include <syslog.h>
int allow_severity = LOG_INFO;
int deny_severity = LOG_INFO;
#endif

int drop_privileges(struct mqtt3_config *config);
void handle_sigint(int signal);
void handle_sigusr1(int signal);
void handle_sigusr2(int signal);

struct mosquitto_db *_mosquitto_get_db(void)
{
	return &int_db;
}

/* mosquitto shouldn't run as root.
 * This function will attempt to change to an unprivileged user and group if
 * running as root. The user is given in config->user.
 * Returns 1 on failure (unknown user, setuid/setgid failure)
 * Returns 0 on success.
 * Note that setting config->user to "root" does not produce an error, but it
 * strongly discouraged.
 */
int drop_privileges(struct mqtt3_config *config)
{
#if !defined(__CYGWIN__) && !defined(WIN32)
	struct passwd *pwd;
	char err[256];

	if(geteuid() == 0){
		if(config->user && strcmp(config->user, "root")){
			pwd = getpwnam(config->user);
			if(!pwd){
				_mosquitto_log_printf(NULL, MOSQ_LOG_ERR, "Error: Invalid user '%s'.", config->user);
				return 1;
			}
			if(initgroups(config->user, pwd->pw_gid) == -1){
				strerror_r(errno, err, 256);
				_mosquitto_log_printf(NULL, MOSQ_LOG_ERR, "Error setting groups whilst dropping privileges: %s.", err);
				return 1;
			}
			if(setgid(pwd->pw_gid) == -1){
				strerror_r(errno, err, 256);
				_mosquitto_log_printf(NULL, MOSQ_LOG_ERR, "Error setting gid whilst dropping privileges: %s.", err);
				return 1;
			}
			if(setuid(pwd->pw_uid) == -1){
				strerror_r(errno, err, 256);
				_mosquitto_log_printf(NULL, MOSQ_LOG_ERR, "Error setting uid whilst dropping privileges: %s.", err);
				return 1;
			}
		}
		if(geteuid() == 0 || getegid() == 0){
			_mosquitto_log_printf(NULL, MOSQ_LOG_WARNING, "Warning: Mosquitto should not be run as root/administrator.");
		}
	}
#endif
	return MOSQ_ERR_SUCCESS;
}

#ifdef SIGHUP
/* Signal handler for SIGHUP - flag a config reload. */
void handle_sighup(int signal)
{
	flag_reload = true;
}
#endif

/* Signal handler for SIGINT and SIGTERM - just stop gracefully. */
void handle_sigint(int signal)
{
	run = 0;
}

/* Signal handler for SIGUSR1 - backup the db. */
void handle_sigusr1(int signal)
{
#ifdef WITH_PERSISTENCE
	flag_db_backup = true;
#endif
}

/* Signal handler for SIGUSR2 - vacuum the db. */
void handle_sigusr2(int signal)
{
	flag_tree_print = true;
}

int main(int argc, char *argv[])
{
	int *listensock = NULL;
	int listensock_count = 0;
	int listensock_index = 0;
	struct mqtt3_config config;
	char buf[1024];
	int i, j;
	FILE *pid;
	int listener_max;
	int rc;
	char err[256];
#ifdef WIN32
	SYSTEMTIME st;
#else
	struct timeval tv;
#endif

#if defined(WIN32) || defined(__CYGWIN__)
	if(argc == 2){
		if(!strcmp(argv[1], "run")){
			service_run();
			return 0;
		}else if(!strcmp(argv[1], "install")){
			service_install();
			return 0;
		}else if(!strcmp(argv[1], "uninstall")){
			service_uninstall();
			return 0;
		}
	}
#endif


#ifdef WIN32
	GetSystemTime(&st);
	srand(st.wSecond + st.wMilliseconds);
#else
	gettimeofday(&tv, NULL);
	srand(tv.tv_sec + tv.tv_usec);
#endif

	memset(&int_db, 0, sizeof(struct mosquitto_db));

	_mosquitto_net_init();

	/*
	 * x.1) initialize the configurations structure by parsing the arguments
	 * passed to mosquitto
	 */
	mqtt3_config_init(&config);
	rc = mqtt3_config_parse_args(&config, argc, argv);

	if(rc != MOSQ_ERR_SUCCESS) return rc;

	int_db.config = &config;

	if(config.daemon){
#ifndef WIN32
		switch(fork()){
			case 0:
				break;
			case -1:
				strerror_r(errno, err, 256);
			_mosquitto_log_printf(NULL, MOSQ_LOG_ERR, "Error in fork: %s", err);
				return 1;
			default:
				return MOSQ_ERR_SUCCESS;
		}
#else
		_mosquitto_log_printf(NULL, MOSQ_LOG_WARNING, "Warning: Can't start in daemon mode in Windows.");
#endif
	}

	if(config.daemon && config.pid_file){
		pid = _mosquitto_fopen(config.pid_file, "wt");
		if(pid){
			fprintf(pid, "%d", getpid());
			fclose(pid);
		}else{
			_mosquitto_log_printf(NULL, MOSQ_LOG_ERR, "Error: Unable to write pid file.");
			return 1;
		}
	}
	rc = drop_privileges(&config);
	if(rc != MOSQ_ERR_SUCCESS) return rc;

	rc = mqtt3_db_open(&config, &int_db);
	if(rc != MOSQ_ERR_SUCCESS){
		_mosquitto_log_printf(NULL, MOSQ_LOG_ERR, "Error: Couldn't open database.");
		return rc;
	}

	/* Initialize logging only after initializing the database in case we're
	 * logging to topics */
	mqtt3_log_init(config.log_type, config.log_dest);
	_mosquitto_log_printf(NULL, MOSQ_LOG_INFO, "mosquitto version %s (build date %s) starting", VERSION, TIMESTAMP);
	if(config.config_file){
		_mosquitto_log_printf(NULL, MOSQ_LOG_INFO, "Config loaded from %s.", config.config_file);
	}else{
		_mosquitto_log_printf(NULL, MOSQ_LOG_INFO, "Using default config.");
	}

//	rc = mosquitto_security_module_init(&int_db);
//	if(rc) return rc;
//	rc = mosquitto_security_init(&int_db, false);
//	if(rc) return rc;

#ifdef WITH_SYS_TREE
	if(config.sys_interval > 0){
		/* Set static $SYS messages */
		snprintf(buf, 1024, "mosquitto version %s", VERSION);
		mqtt3_db_messages_easy_queue(&int_db, NULL, "$SYS/broker/version", 2, strlen(buf), buf, 1);
		snprintf(buf, 1024, "%s", TIMESTAMP);
		mqtt3_db_messages_easy_queue(&int_db, NULL, "$SYS/broker/timestamp", 2, strlen(buf), buf, 1);
#ifdef CHANGESET
		snprintf(buf, 1024, "%s", CHANGESET);
		mqtt3_db_messages_easy_queue(&int_db, NULL, "$SYS/broker/changeset", 2, strlen(buf), buf, 1);
#endif
	}
#endif

	/*
	 * x) initialize mosquitto listeners: 'listeners' control the network ports
	 * at which the mosquitto broker listens for client connections
	 */
	listener_max = -1;
	listensock_index = 0;

	/*
	 * x.1) cycle through the list of listener options, specified in the
	 * configuration files for mosquitto (.conf). listeners are added to
	 *
	 */
	for(i = 0; i < config.listener_count; i++) {

		/*
		 * x.1.1) this function creates a socket and makes it listen on the
		 * port specified on the listener struct. if successful, this function
		 * should return 0.
		 */
		_mosquitto_log_printf(
				NULL,
				MOSQ_LOG_INFO,
				"Initializing listening socket for hostname %s",
				config.listeners[i].host);

		if (xmqtt3_socket_listen(&config.listeners[i])) {
			
			_mosquitto_free(int_db.contexts);
			mqtt3_db_close(&int_db);
			
			if (config.pid_file) {
				remove(config.pid_file);
			}
			
			return 1;
		}

		/*
		 * x.1.2) why do we keep track of the number of sockets used by the
		 * listeners? where is this number set by the way (increment by +1 in
		 * mqtt3_socket_listen())?
		 *
		 * regarding the 1st question, listensock basically consists in an
		 * array of fds, to later initialize an array of pollfd structs, which
		 * will be monitored by the poll() call
		 */
		listensock_count += config.listeners[i].sock_count;
		listensock = (int *) _mosquitto_realloc(listensock, sizeof(int)*listensock_count);
		if(!listensock){
			_mosquitto_free(int_db.contexts);
			mqtt3_db_close(&int_db);
			if(config.pid_file){
				remove(config.pid_file);
			}
			return 1;
		}

		for(j=0; j<config.listeners[i].sock_count; j++){
			if(config.listeners[i].socks[j] == INVALID_SOCKET){
				_mosquitto_free(int_db.contexts);
				mqtt3_db_close(&int_db);
				if(config.pid_file){
					remove(config.pid_file);
				}
				return 1;
			}
			listensock[listensock_index] = config.listeners[i].socks[j];
			if(listensock[listensock_index] > listener_max){
				listener_max = listensock[listensock_index];
			}
			listensock_index++;
		}
	}

	signal(SIGINT, handle_sigint);
	signal(SIGTERM, handle_sigint);
#ifdef SIGHUP
	signal(SIGHUP, handle_sighup);
#endif
#ifndef WIN32
	signal(SIGUSR1, handle_sigusr1);
	signal(SIGUSR2, handle_sigusr2);
	signal(SIGPIPE, SIG_IGN);
#endif

#ifdef WITH_BRIDGE

	// so apparently, 1 broker can have > 1 bridges, i.e. config.bridge_count.
	// here we create multiple bridges.
	for (i = 0; i < config.bridge_count; i++) {

		if (mqtt3_bridge_new(&int_db, &(config.bridges[i]))) {

			_mosquitto_log_printf(
					NULL,
					MOSQ_LOG_WARNING,
					"Warning: Unable to connect to bridge %s.",
					config.bridges[i].name);
		}
	}
#endif

	/*
	 * x.x) main mosquitto loop starts here, after this just cleanup procedures
	 * occur. everything is now hidden under mosquitto_main_loop, in loop.c.
	 */
	run = 1;
	rc = mosquitto_main_loop(
			&int_db,
			listensock,
			listensock_count,
			listener_max);

	_mosquitto_log_printf(
			NULL,
			MOSQ_LOG_INFO,
			"mosquitto version %s terminating",
			VERSION);

	mqtt3_log_close();

#ifdef WITH_PERSISTENCE
	if(config.persistence){
		mqtt3_db_backup(&int_db, true, true);
	}
#endif

	for(i=0; i<int_db.context_count; i++){
		if(int_db.contexts[i]){
			mqtt3_context_cleanup(&int_db, int_db.contexts[i], true);
		}
	}
	_mosquitto_free(int_db.contexts);
	int_db.contexts = NULL;
	mqtt3_db_close(&int_db);

	if(listensock){
		for(i=0; i<listensock_count; i++){
			if(listensock[i] != INVALID_SOCKET){
#ifndef WIN32
				close(listensock[i]);
#else
				closesocket(listensock[i]);
#endif
			}
		}
		_mosquitto_free(listensock);
	}

	mosquitto_security_module_cleanup(&int_db);

	if(config.pid_file){
		remove(config.pid_file);
	}

	_mosquitto_net_cleanup();
	mqtt3_config_cleanup(int_db.config);

	return rc;
}

#ifdef WIN32
int __stdcall WinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance, LPSTR lpCmdLine, int nCmdShow)
{
	char **argv;
	int argc = 1;
	char *token;
	char *saveptr = NULL;
	int rc;

	argv = _mosquitto_malloc(sizeof(char *)*1);
	argv[0] = "mosquitto";
	token = strtok_r(lpCmdLine, " ", &saveptr);
	while(token){
		argc++;
		argv = _mosquitto_realloc(argv, sizeof(char *)*argc);
		if(!argv){
			fprintf(stderr, "Error: Out of memory.\n");
			return MOSQ_ERR_NOMEM;
		}
		argv[argc-1] = token;
		token = strtok_r(NULL, " ", &saveptr);
	}
	rc = main(argc, argv);
	_mosquitto_free(argv);
	return rc;
}
#endif
